lib/OpenQA/WebAPI/GruJob.pm

```diff
@@ -49,8 +50,23 @@ sub execute {
           && exists $self->info->{notes}{gru_id};
     }
     else {
+        # here we call finish, which will either actually finish the job or
+        # retry it
         $self->finish(defined $buffer ? $buffer : 'Job successfully executed');
-        $self->_delete_gru($self->info->{notes}{gru_id}) if exists $self->info->{notes}{gru_id};
+        my $state = $self->info->{state};
+        my $jobid = $self->id;
+        my $gruid;
+        $gruid = $self->info->{notes}{gru_id} if exists $self->info->{notes}{gru_id};
+        log_debug("After execution and finish of minion job $jobid / gru task $gruid, state is $state");
+        # if the job is retried at this point, its state will be 'inactive'
+        # and we do *not* want to delete the corresponding entry from the gru
+        # task database. We only want to delete that if the job is really done
+        if (grep { /$state/ } ('failed', 'finished')) {
+            if ($gruid) {
```

Comment:
in case you're wondering why I nested this: `if (grep { /$state/ } ('failed', 'finished') && $gruid)` doesn't actually seem to *work* right for some reason. If anyone knows why and can fix it, let me know. :D

Reply:
My guess would be that due to precedence rules the last expression (`('failed', 'finished') && $gruid`) gets evaluated first.

Reply:
`if ((grep { /$state/ } ('failed', 'finished')) && $gruid)` should work though, right?

Reply:
Although `grep { $_ eq $state }` would be better style.

Reply:
Actually, why not be totally clear with `if (($state eq 'finished' || $state eq 'failed') && $gruid)`

Reply:
I prefer the last version.

Reply:
ahhh, my old adversary the bracket, it is you once more?!

thanks guys, will tweak.